streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ValidClientsByTaskLoadQueue.java

```diff
@@ -0,0 +1,110 @@
+    private final PriorityQueue<UUID> clientsByTaskLoad;
+    private final BiFunction<UUID, TaskId, Boolean> validClientCriteria;
+
+    ValidClientsByTaskLoadQueue(final Map<UUID, ClientState> clientStates,
+                                final BiFunction<UUID, TaskId, Boolean> validClientCriteria) {
+        clientsByTaskLoad = getClientPriorityQueueByTaskLoad(clientStates);
+        this.validClientCriteria = validClientCriteria;
+    }
+
+    /**
+=     * @return the next least loaded client that satisfies the given criteria, or null if none do
+     */
+    UUID poll(final TaskId task) {
+        final List<UUID> validClient = poll(task, 1);
+        return validClient.isEmpty() ? null : validClient.get(0);
+    }
+
+    /**
+     * @return the next N <= {@code numClientsPerTask} clients in the underlying priority queue that are valid
+     * candidates for the given task
+     */
+    List<UUID> poll(final TaskId task, final int numClients) {
+        final List<UUID> nextLeastLoadedValidClients = new LinkedList<>();
+        final Set<UUID> invalidPolledClients = new HashSet<>();
+        while (nextLeastLoadedValidClients.size() < numClients) {
+            UUID candidateClient;
+            while (true) {
+                candidateClient = clientsByTaskLoad.poll();
+                if (candidateClient == null) {
+                    returnPolledClientsToQueue(invalidPolledClients);
+                    return nextLeastLoadedValidClients;
+                }
+
+                if (validClientCriteria.apply(candidateClient, task)) {
+                    nextLeastLoadedValidClients.add(candidateClient);
+                    break;
+                } else {
+                    invalidPolledClients.add(candidateClient);
+                }
+            }
+        }
+        returnPolledClientsToQueue(invalidPolledClients);
+        return nextLeastLoadedValidClients;
+    }
+
+    void offerAll(final Collection<UUID> clients) {
+        returnPolledClientsToQueue(clients);
+    }
+
+    void offer(final UUID client) {
+        clientsByTaskLoad.offer(client);
+    }
+
+    private void returnPolledClientsToQueue(final Collection<UUID> polledClients) {
+        for (final UUID client : polledClients) {
+            clientsByTaskLoad.offer(client);
+        }
+    }
+
+    static PriorityQueue<UUID> getClientPriorityQueueByTaskLoad(final Map<UUID, ClientState> clientStates) {
+        final PriorityQueue<UUID> queue = new PriorityQueue<>(
+            (client, other) -> {
+                final double clientTaskLoad = clientStates.get(client).taskLoad();
+                final double otherTaskLoad = clientStates.get(other).taskLoad();
+                if (clientTaskLoad < otherTaskLoad) {
+                    return -1;
+                } else if (clientTaskLoad > otherTaskLoad) {
+                    return 1;
+                } else {
+                    return client.compareTo(other);
+                }
+            });
```

Comment:
```suggestion
        final PriorityQueue<UUID> queue = new PriorityQueue<>(
            Comparator.comparingDouble(k -> clientStates.get(k).taskLoad())
        );
```

Upon second reading, this does the same thing, right?

Reply:
Almost, we want to fall back to comparing the actual UUIDs if the `taskLoad` happens to be equal

Reply:
Ah, thanks.